nativelib/src/main/resources/gc/commix/Allocator.c

```diff
@@ -0,0 +1,427 @@
+ * recyclable block for normal allocation.
+ *
+ * @param allocator
+ * @return `true` if there are enough block to initialise the cursors, `false`
+ * otherwise.
+ */
+bool Allocator_CanInitCursors(Allocator *allocator) {
+    uint32_t freeBlockCount =
+        (uint32_t)allocator->blockAllocator->freeBlockCount;
+    return freeBlockCount >= 2 ||
+           (freeBlockCount == 1 && allocator->recycledBlockCount > 0);
+}
+
+void Allocator_Clear(Allocator *allocator) {
+    BlockList_Clear(&allocator->recycledBlocks);
+    allocator->recycledBlockCount = 0;
+    allocator->limit = NULL;
+    allocator->block = NULL;
+    allocator->largeLimit = NULL;
+    allocator->largeBlock = NULL;
+}
+
+bool Allocator_newOverflowBlock(Allocator *allocator) {
+    BlockMeta *largeBlock =
+        BlockAllocator_GetFreeBlock(allocator->blockAllocator);
+    if (largeBlock == NULL) {
+        return false;
+    }
+    allocator->largeBlock = largeBlock;
+    word_t *largeBlockStart = BlockMeta_GetBlockStart(
+        allocator->blockMetaStart, allocator->heapStart, largeBlock);
+    allocator->largeBlockStart = largeBlockStart;
+    allocator->largeCursor = largeBlockStart;
+    allocator->largeLimit = Block_GetBlockEnd(largeBlockStart);
+    return true;
+}
+
+/**
+ * Overflow allocation uses only free blocks, it is used when the bump limit of
+ * the fast allocator is too small to fit
+ * the block to alloc.
+ */
+word_t *Allocator_overflowAllocation(Allocator *allocator, size_t size) {
+    word_t *start = allocator->largeCursor;
+    word_t *end = (word_t *)((uint8_t *)start + size);
+
+    // allocator->largeLimit == NULL implies end > allocator->largeLimit
+    assert(allocator->largeLimit != NULL || end > allocator->largeLimit);
+    if (end > allocator->largeLimit) {
+        if (!Allocator_newOverflowBlock(allocator)) {
+            return NULL;
+        }
+        return Allocator_overflowAllocation(allocator, size);
+    }
+
+    memset(start, 0, size);
+
+    allocator->largeCursor = end;
+
+    return start;
+}
+
+/**
+ * Allocation fast path, uses the cursor and limit.
+ */
+INLINE word_t *Allocator_tryAlloc(Allocator *allocator, size_t size) {
+    word_t *start = allocator->cursor;
+    word_t *end = (word_t *)((uint8_t *)start + size);
+
+    // allocator->limit == NULL implies end > allocator->limit
+    assert(allocator->limit != NULL || end > allocator->limit);
+    if (end > allocator->limit) {
+        // If it overlaps but the block to allocate is a `medium` sized block,
+        // use overflow allocation
+        if (size > LINE_SIZE) {
+            return Allocator_overflowAllocation(allocator, size);
+        } else {
+            // Otherwise try to get a new line.
+            if (Allocator_getNextLine(allocator)) {
+                return Allocator_tryAlloc(allocator, size);
+            }
+
+            return NULL;
+        }
+    }
+
+    memset(start, 0, size);
```

Comment:
Why do we need memset here? `tryAlloc` is only called on allocation slow path that returns back to fast path before memory is zeroed.

Reply:
When the slow path is called the object is returned immediately and the memset in fast path is not called. However I did manage to reduce the call instructions to memset by moving them from `tryAlloc` and `Allocator_overflowAllocation` to `Allocator_allocSlow`.  `tryAlloc` gets inlined in multiple places, so this should reduce the code size somewhat.